packages/frontend/src/features/scheduler/components/SchedulerFilters.tsx

```diff
@@ -100,51 +100,86 @@ const FilterItem: FC<SchedulerFilterItemProps> = ({
+    );
+    const originalDashboardFilters = dashboard?.filters;
+
+    const [schedulerFiltersData, setSchedulerFiltersData] = useState<
+        DashboardFilterRule[] | undefined
+    >(schedulerFilters);
 
     const handleUpdateSchedulerFilter = useCallback(
         (schedulerFilter: SchedulerFilterRule) => {
-            // TODO: this should diff if the filter is actually
-            // different from the dashboard filter
-
-            const newState = produce(schedulerFilters, (draft) => {
-                const filterIndex = draft.findIndex(
-                    (f) =>
-                        f.target.fieldId === schedulerFilter.target.fieldId &&
-                        f.target.tableName === schedulerFilter.target.tableName,
-                );
+            if (!originalDashboardFilters) return;
 
-                if (draft[filterIndex]) {
-                    draft[filterIndex] = schedulerFilter;
-                } else {
-                    draft.push(schedulerFilter);
-                }
-            });
+            const originalFilter = originalDashboardFilters.dimensions.find(
+                (d) => d.id === schedulerFilter.id,
+            );
 
-            setSchedulerFilters(newState);
+            if (!originalFilter) return;
 
-            // TODO: sync with upper component
-            // call onChange - this can be debounced
-        },
-        [schedulerFilters],
-    );
+            let updatedFilters = schedulerFiltersData
+                ? [...schedulerFiltersData]
+                : [];
 
-    const isLoadingDashboardFilters = useDashboardContext(
-        (c) => c.isLoadingDashboardFilters,
-    );
-    const allFilters = useDashboardContext((c) => c.allFilters);
-    const fieldsWithSuggestions = useDashboardContext(
-        (c) => c.fieldsWithSuggestions,
+            const filterIndex = updatedFilters.findIndex(
+                (f) => f.id === schedulerFilter.id,
+            );
+            const isExistingFilter = filterIndex !== -1;
+
+            const filterToCompareAgainst = isExistingFilter
+                ? updatedFilters[filterIndex]
+                : originalFilter;
+
+            const { tileTargets, ...schedulerFilterWithoutTileTargets } =
+                schedulerFilter;
+
+            const hasChanged = hasSavedFilterValueChanged(
+                filterToCompareAgainst,
+                schedulerFilterWithoutTileTargets,
+            );
+
+            const isReverted = !hasSavedFilterValueChanged(
+                originalFilter,
+                schedulerFilterWithoutTileTargets,
+            );
+
+            if (hasChanged) {
+                if (isExistingFilter) {
+                    updatedFilters[filterIndex] = schedulerFilter;
+                } else {
+                    updatedFilters.push(schedulerFilter);
+                }
+            }
+            if (isReverted) {
+                updatedFilters.splice(filterIndex, 1);
+            }
```

Comment:
Note: found it hard to review this logic because we are creating variables and making calculations ahead of time. I had to go back and forward in the code to recreate all the different scenarios. I still don't know if there is an edge case where both `hasChanged` and `isReverted` can be `true`


eg: if we know they reverted the filter value, we don't need the `filterToCompareAgainst`, `hasChanged` etc

I wonder if we could have a `if else` where all the logic for `hasChanged` would be inside the else block

Reply:
Yea, agree. This is def not that straightforward. I've separated this logic into 2 functions: one that handles the `onChange` and a middle one for updating the fields and getting the correct types + updating/removing the changes. 
Also, the check if `isReverted` has been moved further up and it returns so that we avoid extra calculations when that's the case. 

I'm also not sure if there is such a case: reverted **and** changed. I think it's not possible atm